walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,75 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.ListIterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      if (request.getAttendees().size() == 0)
+          return performQuery(events, request, true);
+      Collection<TimeRange> freeList = performQuery(events, request, true);
+      if (freeList.size() == 0)
+          return performQuery(events, request, false);
+      return freeList;
+  }
+  
+  public Collection<TimeRange> performQuery(Collection<Event> events, MeetingRequest request, boolean considerOptionalAttendees) {
+    LinkedList<TimeRange> freeList = new LinkedList<TimeRange>();
+    freeList.add(TimeRange.WHOLE_DAY);
+
+    // Carve freeList down into series of non-conflicting chunks
+    for (Event event : events) {
+        if (( !considerOptionalAttendees
+              || Collections.disjoint(event.getAttendees(), request.getOptionalAttendees()))
+        && Collections.disjoint(event.getAttendees(), request.getAttendees()))
+            continue;
+
+        int eventStart = event.getWhen().start();
+        int eventEnd = event.getWhen().end();
+        ListIterator<TimeRange> iter = freeList.listIterator(0);
+
+        while (iter.hasNext()) {
+            TimeRange freeBlock = iter.next();
+            if (eventEnd <= freeBlock.start())
+                break;
+            else if (eventStart < freeBlock.end())
+                trim(iter, freeBlock, eventStart, eventEnd);
+        }
+    }
+
+    // Trim off freeBlocks of insufficient size
+    long requiredDuration = request.getDuration();
+    ListIterator<TimeRange> iter = freeList.listIterator(0);
+    while (iter.hasNext()) {
+        TimeRange freeBlock = iter.next();
+        if ((long) freeBlock.duration() < requiredDuration)
+            iter.remove();
+    }
+
+    return(freeList);
+  }
+
+  private void trim(ListIterator iter, TimeRange freeBlock, int eventStart, int eventEnd) {
+      if (eventStart <= freeBlock.start() && eventEnd >= freeBlock.end())   // Event fully contains free-block
+          iter.remove();
+      else if (eventStart <= freeBlock.start()) {   // Event overlaps with beginning of free-block
```

Comment:
One simple measurement how good your test cases are is called "line coverage". It is essentially the percentage of lines of code your tests executes. We generally try to get 100% line coverage - i.e. every line that contains some kind of logic is executed in at least one of the test cases. Correct me if I'm wrong, but I don't think you have a test case where an event partially overlap a free-block. I think this only happens when existing events partially overlap each other.

Reply:
I was sure that was covered somewhere, but I guess I need to read better. I'll be sure this is added!